device/device_test.go

```diff
@@ -474,3 +475,37 @@ func TestBatchSize(t *testing.T) {
 		t.Errorf("expected batch size %d, got %d", want, got)
 	}
 }
+
+var packetSink []byte
+
+func BenchmarkMessageInitiationMarshal(b *testing.B) {
+	var msg MessageInitiation
+
+	b.Run("binary.Write", func(b *testing.B) {
+		b.ReportAllocs()
+		for range b.N {
+			var buf [MessageInitiationSize]byte
+			writer := bytes.NewBuffer(buf[:0])
+			_ = binary.Write(writer, binary.LittleEndian, msg)
+			packetSink = writer.Bytes()
+		}
+	})
+
+	b.Run("binary.Encode", func(b *testing.B) {
+		b.ReportAllocs()
+		for range b.N {
+			packet := make([]byte, MessageInitiationSize)
+			_, _ = binary.Encode(packet, binary.LittleEndian, msg)
+			packetSink = packet
+		}
+	})
```

Comment:
Note: https://pkg.go.dev/encoding/binary#Encode was added in go 1.23, its better than binary.Write but of course can't beat plain copy.

Reply:
What's the performance difference like?

Reply:
See benchmark commit